edx_notifications/base_data.py

```diff
@@ -0,0 +1,286 @@
+"""
+Base objects that data.py uses
+"""
+
+import json
+import copy
+import inspect
+import dateutil.parser
+
+from datetime import datetime
+
+from weakref import WeakKeyDictionary
+
+
+class TypedField(object):
+    """
+    Field Decscriptors used to enforce correct typing
+    """
+
+    _expected_type = None
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initializer which takes in the type this field
+        should be set it is set
+        """
+
+        self._expected_type = args[0]
+        self._default = kwargs.get('default', None)
+        self._data = WeakKeyDictionary()
```

Comment:
Why store the data for each instance on the field object, rather than storing it on the instance objects?